Sort modifiers

jhaber opened the pull request:
Related to #291 

I had some trouble getting access to the actual modifier name ("public", "static", "private", etc.) and I don't write much javascript so any feedback on the code is welcome. 

There is an existing `sortModifiers` function that everything goes through, however it seems to just separate leading annotations from all other annotations/modifiers. I tried to preserve the behavior of this function, however it conflicts a bit with the new modifier sorting behavior. 

For example, if you have this code:
`final @Annotation private String field;`

Currently, prettier-java leaves this code alone. However, we want to swap the order of the `final` and `private` modifiers, but it's not clear what should happen to `@Annotation` when we do this. As currently written, this PR moves all such annotations to the end, and would output:
`private final @Annotation String field;`

Some other options that come to mind:
- unconditionally move annotations ahead of modifiers, which means you can't write a method signature like `public @Nullable String myMethod()`
- move annotations to the front unless they come after all modifiers, which still allows method signatures like `public @Nullable String myMethod()`

clement26695 commented:
Just maybe one thing to add:would it be possible to add a docs section in the prettier-plugin-java package with a description of this sorting behavior as it might not seem directly intuitive?

jhaber commented:
Sure thing, is there an existing place where that sort of documentation should go or should I add a docs folder within `packages/prettier-plugin-java`?

clement26695 commented:
No, there is no existing place yet. Can you create the folder please?

jhaber commented:
I added `packages/prettier-plugin-java/docs/modifiers.md`, let me know how it looks to you

clement26695 commented:
LGTM! I will have a final review before merging it but it should be good

murdos commented:
> 
> 
> LGTM. I don't think there is a better way for accessing the type of modifier. I'm not sure either about what should be done with annotations but this one looks good to me. Do you have an opinion about this @murdos?

For a field I would expect `final @Annotation private String field;` to become
```java
@Annotation
private final String field;
```

but as @jhaber said, for a method, we should correctly handle annotation attached to return type of a method : `public @Nullable String myMethod()`

jhaber commented:
Yeah, thinking about it more the concern with preserving annotation placement only applies to methods where you have ambiguity about method annotations vs. return type annotations. So I would propose updating the logic so that annotations unconditionally get moved to the front and on their own lines for  classes and fields. For methods, the same will happen unless the annotation comes after all other modifiers

jhaber commented:
I've updated the code and documentation to implement the approach described in my previous comment. However, this is causing a test failure on `comments/class/_input.java`:
```java
private transient @MonotonicNonNull ColumnMap columnMap;
```

Previously this was being left alone, but is now getting re-written to:
```java
@MonotonicNonNull
private transient ColumnMap columnMap;
```

This matches my expectation, so  I updated `comments/class/_output.java` to make the test pass. Let me know if that's the wrong approach to fixing the issue.

clementdessoude commented:
Thank you again @jhaber !

jhaber commented:
Glad I could help a tiny bit with this awesome project 😄
